extensions/chathistory.md

```diff
@@ -0,0 +1,88 @@
+
+Both the `message_count` and `max_message_count` MUST be non-zero integers and `max_message_count` MUST be positive. The client should not request a `message_count` with an absolute value greater than the `max_message_count` parameter sent in the command. If the `message_count` absolute value exceeds the `max_message_count`, server should return a number of lines equal to the `max_message_count` and the appropriate warning as described below.
+
+#### Format
+The `chathistory` content can requested using timestamps:
+
+    @draft/label=ID CHATHISTORY target timestamp=YYYY-MM-DDThh:mm:ss.sssZ message_count
+
+Alternatively, content can be requested using a `draft/msgid`:
+
+    @draft/label=ID CHATHISTORY target draft/msgid=ID message_count
+
+If `message_count` is positive, content MUST be retrieved from after and not including the specified timestamp or `draft/msgid`. If the `message_count` is negative, content MUST be retrieved from before the specified timestamp or `draft/msgid`.
+
+#### Errors and Warnings
+If the server receives an improperly formatted `CHATHISTORY` command, the `CMD_INVALID` error code should be returned.
+
+If the absolute value of `message_count` exceeds the `max_message_count`, warn code `MAX_MSG_COUNT_EXCEEDED` should be returned. The command should continue to be processed as described above.
+
+If no `chathistory` exists to return, the server should return the appropriate error code. `ACCESS_DENIED` should be sent if the user requests content they do not have permission to view.
+
+### Examples
+The examples below are written with `draft/msgid` and `draft/label` tags included. These tags are recommended.
+
+#### Begin
+    @draft/label=ID :irc.host BATCH +ID chathistory target
+#### PRIVMSG
+    @batch=ID;draft/msgid=ID;time=YYYY-MM-DDThh:mm:ss.sssZ :nick!ident@host PRIVMSG target :message
+#### NOTICE
+    @batch=ID;draft/msgid=ID;time=YYYY-MM-DDThh:mm:ss.sssZ :nick!ident@host NOTICE target :message
+#### End
+    :irc.host BATCH -ID
+#### Error
+    @draft/label=ID :nick!ident@host CHATHISTORY ERR ERROR_CODE
+#### Warning
+    @draft/label=ID :nick!ident@host CHATHISTORY WARN WARN_CODE
+
+## Use Cases
+The batch type and supporting command are useful for allowing an "infinite scroll" type capability within the client. A client will, upon scrolling to the top of the active window or a manual trigger, may request `chathistory` from the server and, after receiving returned content, append it to the top of the window. Users can repeat this historic scrolling to retrieve prior history until limitations are met (see below).
```

Comment:
> If `message_count` is positive, content MUST be retrieved from after and not including the specified timestamp or `draft/msgid`. If the `message_count` is negative, content MUST be retrieved from before the specified timestamp or `draft/msgid`.

The client is sending a known message as a starting point. If the oldest message a client has is at a particular time, this timestamp will be provided to `CHATHISTORY` and anything before it will be returned. The same goes for `draft/msgid`.

This works in the forward direction as well, in which the client knows the last message it has and wants to retrieve content from directly after it.

Theoretically, this works for any known message as a reference, and could be used to add context if a client wants to "jump" to a message, and then request X lines before and Y lines after.